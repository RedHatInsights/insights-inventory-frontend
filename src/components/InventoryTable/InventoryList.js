/* eslint-disable react/display-name */
import React, { useEffect, useRef } from 'react';
import { useSelector } from 'react-redux';
import InventoryEntityTable from './EntityTable';
import { Grid, GridItem } from '@patternfly/react-core';
import PropTypes from 'prop-types';
import './InventoryList.scss';
import isEqual from 'lodash/isEqual';
import AccessDenied from '../../Utilities/AccessDenied';

const convertItem = ({ children, isOpen, ...item }) => item;

/**
 * Component that works as a side channel for consumers to notify inventory of new data changes.
 *  @param   {object}     props               props object
 *  @param   {boolean}    props.showHealth    if true, the health is shown
 *  @param   {Function}   props.onRefreshData on refresh data function to notify inventory of new data changes
 *  @param   {boolean}    props.ignoreRefresh if true, the ignore refresh is enabled and the onRefreshData function is not called
 *  @returns {React.node}                     React node with inventory list
 */
const ContextInventoryList = ({
  showHealth,
  onRefreshData,
  ignoreRefresh,
  ...props
}) => {
  const prevItems = useRef(props.items);
  const prevSortBy = useRef(props.sortBy);

  /* eslint-disable react-hooks/exhaustive-deps */
  useEffect(() => {
    if (props.hasItems) {
      onRefreshData({}, ignoreRefresh);
    }
  }, []);
  /* eslint-enable react-hooks/exhaustive-deps */

  /**
   * Function to calculate for new changes, this function limits re-renders by checking if previous items are
   * same as new items.
   * If items are not passed, it only checks for props sortBy.
   *  @param   {object} prevProps previous props - items, hasItems, sortBy.
   *  @returns {void}             void
   */
  useEffect(() => {
    if (
      props.hasItems &&
      !isEqual(prevItems.current.map(convertItem), props.items.map(convertItem))
    ) {
      prevItems.current = props.items;
      onRefreshData({}, ignoreRefresh);
    }

    if (
      !props.hasItems &&
      props.loaded == true &&
      !isEqual(prevSortBy.current, props.sortBy)
    ) {
      prevSortBy.current = props.sortBy;
      onRefreshData({});
    }
  });

  return (
    <Grid
      gutter="sm"
      className="ins-inventory-list"
      data-testid="inventory-table-list"
    >
      <GridItem span={12}>
        <InventoryEntityTable {...props} onRefreshData={onRefreshData} />
      </GridItem>
    </Grid>
  );
};

/**
 * Component that consumes active filters and passes them down to component.
 *  @param   {object}     props               props object
 *  @param   {boolean}    props.hasAccess     if true, the access is granted
 *  @param   {Function}   props.onRefreshData on refresh data function to notify inventory of new data changes
 *  @returns {React.node}                     React node with inventory list
 */
const InventoryList = React.forwardRef(
  ({ hasAccess, onRefreshData, ...props }, ref) => {
    const activeFilters = useSelector(
      ({ entities: { activeFilters } }) => activeFilters,
    );

    if (ref) {
      ref.current = {
        onRefreshData: (params, disableRefresh = true, forceRefresh) =>
          onRefreshData(params, disableRefresh, forceRefresh),
      };
    }

    return !hasAccess ? (
      <div className="ins-c-inventory__no-access">
        <AccessDenied showReturnButton={false} />
      </div>
    ) : (
      <ContextInventoryList
        {...props}
        activeFilters={activeFilters}
        onRefreshData={onRefreshData}
      />
    );
  },
);

ContextInventoryList.propTypes = {
  ...InventoryList.propTypes,
  setRefresh: PropTypes.func,
  onRefreshData: PropTypes.func,
  ignoreRefresh: PropTypes.bool,
};
ContextInventoryList.defaultProps = {
  perPage: 50,
  page: 1,
  ignoreRefresh: true,
};
InventoryList.propTypes = {
  showTags: PropTypes.bool,
  filterEntities: PropTypes.func,
  loadEntities: PropTypes.func,
  showHealth: PropTypes.bool,
  page: PropTypes.number,
  perPage: PropTypes.number,
  sortBy: PropTypes.shape({
    key: PropTypes.string,
    direction: PropTypes.string,
  }),
  items: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.shape({
        id: PropTypes.string.isRequired,
      }),
      PropTypes.shape({
        account: PropTypes.any,
        isOpen: PropTypes.bool,
        title: PropTypes.node,
      }),
    ]),
  ),
  entities: PropTypes.arrayOf(PropTypes.any),
  customFilters: PropTypes.shape({
    tags: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.arrayOf(PropTypes.string),
    ]),
  }),
  getEntities: PropTypes.func,
  hideFilters: PropTypes.shape({
    tags: PropTypes.bool,
    name: PropTypes.bool,
    registeredWith: PropTypes.bool,
    stale: PropTypes.bool,
    operatingSystem: PropTypes.bool,
  }),
  onRefreshData: PropTypes.func,
};

InventoryList.defaultProps = {
  hasAccess: true,
};

export default InventoryList;
