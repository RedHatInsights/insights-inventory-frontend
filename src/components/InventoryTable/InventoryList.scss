.ins-c-inventory__no--access .pf-c-empty-state__body > div {
    max-width: 500px;
    margin: auto;

    @media screen and (max-width: 768px) {
        max-width: 200px;
    }
}

.ins-c-inventory__table--toolbar {
    &.ins-m-footer {
        margin: 0;
        padding: 0;
    }

    &:not(.ins-c-inventory__table--toolbar-has-items) {
        .pf-c-toolbar__content {
            .pf-c-toolbar__item:last-child { min-width: 12.5rem; }
        }
        .pf-c-pagination__nav-page-select input {  min-width: 4.68755rem; }
    }

    .ins-c-chip-filters .pf-c-chip-group:not(:last-of-type) { margin-right: var(--pf-global--spacer--xs); }

    .ins-c-tagfilter { width: initial; }
}

.ins-c-table__empty {
    margin: 4.68755rem;
    border-bottom: none;

    .pf-c-content {
        text-align: center;
    }
}

.ins-c-inventory__tags-filter ul {
    label.pf-c-check__label {
        display: flex;
        width: 100%;

        .ins-c-inventory__tags-filter--tag-name {
            margin-right: var(--pf-global--spacer--xs);
        }
        .pf-c-badge { margin-left: auto; }
    }

    .ins-c-inventory__tags-tail {
        display: flex;

        .ins-c-inventory__tags-no-tags { margin: auto; }

        .pf-c-spinner { margin-left: auto; }
    }
}

.ins-simple-table {
    .ins-health-status div {
        display: inline-block;
        padding-right: 20px;

        svg {
          position: relative;
            top: 2px;
            left: -2px;
        }
    }
}

.ins-c-entity-table.pf-c-table {
    td {
        vertical-align: middle;
    }

    .pf-c-table__check input { cursor: pointer; }
    .ins-health-status div {
        display: inline-block;
        padding-right: 20px;

        svg {
            position: relative;
            top: 2px;
            left: -2px;
        }

        span:nth-child(2) {
            color: var(--pf-global--info-color--100);
            cursor: pointer;
        }
    }

    .ins-composed-col {
        > :first-child {
            color: var(--pf-c-button--m-disabled--Color);
            font-size: var(--pf-global--FontSize--sm);
        }

        * {
            font-size: var(	--pf-global--FontSize--lg);
        }

        .ins-composed-col {
            cursor: pointer;
        }
    }

    .ins-c-inventory__list-tags {
        display: inline-block;
    }

    .pf-c-table__action .pf-c-dropdown {
        position: relative;
        left: 10px;
        top: 3px;
    }

    .ins-composed-col div {
        word-wrap: break-word;
        word-break:break-all;
        &:nth-child(2):not(.ins-m-nodetail) {
            color: var(--pf-global--link--Color);
            cursor: pointer;
            &:hover { color: var(--pf-global--link--Color--hover); }
        }
    }
}

.ins-c-inventory__tags-modal .ins-c-primary-toolbar .pf-c-toolbar__content {
    padding: 0;
}

.ins-c-inventory__no-access { background-color: var(--pf-global--BackgroundColor--100); }
