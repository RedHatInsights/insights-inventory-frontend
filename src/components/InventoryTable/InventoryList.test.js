import { render, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { createPromise as promiseMiddleware } from 'redux-promise-middleware';
import InventoryList from './InventoryList';

jest.mock('../../Utilities/useFeatureFlag.js');
jest.mock('../../store/actions', () => {
  const actions = jest.requireActual('../../store/actions');
  const { ACTION_TYPES } = jest.requireActual('../../store/action-types');
  return {
    __esModule: true,
    ...actions,
    loadEntities: () => ({
      type: ACTION_TYPES.LOAD_ENTITIES,
      payload: () => Promise.resolve([]),
    }),
  };
});

describe('InventoryList', () => {
  let initialState;
  let mockStore;
  let ref;
  let onRefreshData;

  beforeEach(() => {
    ref = { current: undefined };
    mockStore = configureStore([promiseMiddleware()]);
    initialState = {
      entities: {
        activeFilters: [{}],
        loaded: true,
        rows: [],
        columns: [{ key: 'one', title: 'One' }],
        page: 1,
        perPage: 50,
        total: 500,
      },
    };
    onRefreshData = jest
      .fn()
      .mockImplementation(() => Promise.resolve({ results: [], total: 0 }));
  });

  it('should render correctly', () => {
    const store = mockStore(initialState);
    render(
      <MemoryRouter>
        <Provider store={store}>
          <InventoryList ref={ref} onRefreshData={onRefreshData} loaded />
        </Provider>
      </MemoryRouter>
    );

    screen.getByRole('columnheader', {
      name: /one/i,
    });
    screen.getByRole('grid', {
      name: /host inventory/i,
    });
    screen.getByRole('heading', {
      name: /no matching systems found/i,
    });
    screen.getByText(/to continue, edit your filter settings and try again/i);
  });

  it('should render correctly - with no access', () => {
    const store = mockStore(initialState);
    render(
      <MemoryRouter>
        <Provider store={store}>
          <InventoryList
            hasAccess={false}
            ref={ref}
            onRefreshData={onRefreshData}
            loaded
          />
        </Provider>
      </MemoryRouter>
    );

    screen.getByRole('heading', {
      name: /you do not have access to inventory/i,
    });
    screen.getByText(
      /to view your systems, you must be granted inventory access from your organization administrator\./i
    );
  });

  describe('API', () => {
    beforeEach(() => {
      onRefreshData.mockClear();
    });

    it('should fire refresh after changing sort', () => {
      const sortBy = {
        index: 1,
        key: 'one',
        direction: 'asc',
      };
      const store = mockStore(initialState);
      const Cmp = (props) => (
        <MemoryRouter>
          <Provider store={store}>
            <InventoryList
              {...props}
              ref={ref}
              onRefreshData={onRefreshData}
              loaded
            />
          </Provider>
        </MemoryRouter>
      );
      const { rerender } = render(<Cmp sortBy={sortBy} />);

      rerender(
        <Cmp
          sortBy={{
            ...sortBy,
            direction: 'desc',
          }}
        />
      );
      expect(onRefreshData).toHaveBeenCalledTimes(1);
    });

    it('should not fire refresh after changing props', () => {
      const sortBy = {
        index: 1,
        key: 'one',
        direction: 'asc',
      };
      const store = mockStore(initialState);
      const Cmp = (props) => (
        <MemoryRouter>
          <Provider store={store}>
            <InventoryList
              {...props}
              ref={ref}
              onRefreshData={onRefreshData}
              loaded
            />
          </Provider>
        </MemoryRouter>
      );
      const { rerender } = render(<Cmp sortBy={sortBy} />);

      rerender(<Cmp sortBy={sortBy} showTags />);
      expect(onRefreshData).toHaveBeenCalledTimes(0);
    });

    it('should fire refresh after changing items', () => {
      const store = mockStore(initialState);
      const Cmp = (props) => (
        <MemoryRouter>
          <Provider store={store}>
            <InventoryList
              {...props}
              ref={ref}
              onRefreshData={onRefreshData}
              loaded
            />
          </Provider>
        </MemoryRouter>
      );
      const { rerender } = render(
        <Cmp
          items={[
            { children: () => <div>test</div>, isOpen: false, id: 'fff' },
          ]}
          hasItems
        />
      );

      rerender(
        <Cmp
          items={[
            { children: () => <div>test</div>, isOpen: false, id: 'something' },
          ]}
          hasItems
        />
      );
      expect(onRefreshData).toHaveBeenCalledTimes(2);
    });

    it('should fire refresh calling it from ref', () => {
      const store = mockStore(initialState);
      const Cmp = (props) => (
        <MemoryRouter>
          <Provider store={store}>
            <InventoryList
              {...props}
              ref={ref}
              onRefreshData={onRefreshData}
              loaded
            />
          </Provider>
        </MemoryRouter>
      );
      render(
        <Cmp
          items={[
            { children: () => <div>test</div>, isOpen: false, id: 'fff' },
          ]}
          hasItems
        />
      );

      expect(onRefreshData).toHaveBeenCalledTimes(1);
      ref.current.onRefreshData();
      expect(onRefreshData).toHaveBeenCalledTimes(2);
    });
  });
});
