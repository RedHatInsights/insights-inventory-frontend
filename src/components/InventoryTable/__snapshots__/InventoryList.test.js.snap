// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryList should render correctly - with no access 1`] = `
<div
  class="ins-c-inventory__no-access"
>
  <div
    class="pf-c-empty-state ins-c-not-authorized ins-c-inventory__no--access"
  >
    <div
      class="pf-c-empty-state__content"
    >
      <svg
        aria-hidden="true"
        class="pf-c-empty-state__icon"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align:-0.125em"
        viewBox="0 0 448 512"
        width="1em"
      >
        <path
          d="M400 224h-24v-72C376 68.2 307.8 0 224 0S72 68.2 72 152v72H48c-26.5 0-48 21.5-48 48v192c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V272c0-26.5-21.5-48-48-48zm-104 0H152v-72c0-39.7 32.3-72 72-72s72 32.3 72 72v72z"
        />
      </svg>
      <h5
        class="pf-c-title pf-m-lg"
        data-ouia-component-id="OUIA-Generated-Title-1"
        data-ouia-component-type="PF4/Title"
        data-ouia-safe="true"
      >
        You do not have access to Inventory
      </h5>
      <div
        class="pf-c-empty-state__body"
      >
        <div>
          To view your systems, you must be granted inventory access from your Organization Administrator.
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`InventoryList should render correctly 1`] = `
<div
  class="pf-l-grid ins-inventory-list"
  gutter="sm"
>
  <div
    class="pf-l-grid__item pf-m-12-col"
  >
    <table
      aria-label="Host inventory"
      class="ins-c-entity-table pf-c-table pf-m-grid-md pf-m-compact pf-m-sticky-header"
      data-ouia-component-id="OUIA-Generated-Table-2"
      data-ouia-component-type="PF4/Table"
      data-ouia-safe="true"
      role="grid"
    >
      <thead
        class=""
      >
        <tr
          aria-label=""
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-1"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <th
            class=""
            data-key="0"
            data-label="One"
            scope="col"
          >
            One
          </th>
        </tr>
      </thead>
      <tbody
        class=""
        role="rowgroup"
      >
        <tr
          aria-label=""
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-2"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            colspan="1"
            data-key="0"
          >
            <div
              class="pf-c-empty-state"
              data-ouia-component-id="empty-state"
              data-ouia-component-type="PF4/EmptyState"
              data-ouia-safe="true"
            >
              <div
                class="pf-c-empty-state__content"
              >
                <svg
                  aria-hidden="true"
                  class="pf-c-empty-state__icon"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align:-0.125em"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                  />
                </svg>
                <h5
                  class="pf-c-title pf-m-lg"
                  data-ouia-component-id="OUIA-Generated-Title-1"
                  data-ouia-component-type="PF4/Title"
                  data-ouia-safe="true"
                >
                  No matching systems found
                </h5>
                <div
                  class="pf-c-empty-state__body"
                >
                  To continue, edit your filter settings and try again
                </div>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
